docs/design/mp_arktos.md

```diff
@@ -0,0 +1,241 @@
+ Sample Pod definition:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    arktos.futurewei.com/network: vpc-1
+  annotations:
+    arktos.futurewei.com/nic: {"name": "eth0", "ip": "10.10.1.12", "subnet": "net1"}
+    arktos.futurewei.com/vip-hint: {"cidr":"10.10.1.0/26", "ip":"10.10.1.12", "policy":"BestEffort"}
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+    ports:
+      - containerPort: 443
+```
+Operator kicks off creation process:
+* List Kubernetes cluster object / Arktos newtork object
+* Retrieve ```labels``` and ```annotations``` information: 
+    * ```arktos.futurewei.com/network```: vpc-1
+    * ```arktos.futurewei.com/nic```: {"name": "eth0", "ip": "10.10.1.12", "subnet": "net1"}
+* Find Arktos network object that is listed in ```arktos.futurewei.com/network```
+* Find subnet and ip information from ```arktos.futurewei.com/nic```, and choose the correct subnet to place the new pod. 
+  * If subnet is the only specified information, Mizar will launch the Pod in ```net1``` and assign an ip address to the Pod. Please note that Mizar only supports Pod creation in an existing subnet. In other words, if ```net1``` listed in definition does not exist, Mizar will throw an error. 
+  * If only  ```ip``` is specified, Mizar needs to figure out which subnet this ip falls into, and then creates the pod into the correct subnet. 
+  * If both ip and subnet are missing, Mizar will choose the first subnet within ```vpc-1``` to launch the Pod and assign an ip address to that Pod. 
+* Update object store: map subnet object (which is equivalent to net object in Mizar) with pod: ```self.store.eps_net_store[net1][nginx] = pod object```
+* After Pod creation is completed, update Pod object with new annotations: ```mizar.futurewei.com/network_user_input``` and ```arktos.futurewei.com/network-readiness```.      
+* ```mizar.futurewei.com/network_user_input``` will include Mizar network information such as ```vpc name```, ```subnet name``` and ```ip address```
+* In cni daemon service, it will retrieve newly created Pod object, and pass in required network configuration. See below: 
+
+```
+    def cni_add(self, params):
+        ep = self.get_pod_obj(params)
+
+        result = {
+            "cniVersion": params.cni_version,
+            "interfaces": [
+                {
+                    "name": pod.veth_name,
+                    "mac": pod.mac,
+                    "sandbox": pod.netns
+                }
+            ],
+            "ips": [
+                {
+                    "version": "4",
+                    "address": "{}/{}".format(pod.ip, pod.prefix),
+                    "gateway": pod.gw,
+                    "interface": 0
+                }
+            ]
+        }
+
+    def get_pod_obj(self, params):
+        logger.debug("Allocate endpoint name")
+        name = ""
+        if 'K8S_POD_NAME' in params.cni_args_dict:
+            name = params.cni_args_dict['K8S_POD_NAME']
```

Comment:
it seems mizar will use pod-name as the only key. This works fine most of the cases; however it might conflict with other namespace/tenant.

Reply:
I am thinking to use K8S_POD_NAMESPACE as a way to determine the tenant ID. However, since metadata.tenant and metadata.namespace is different in Arktos, what is the best way for Mizar to get tenant information? As I mentioned during our conversation off-line, the arktos network object definition and pod definition do not have tenant information.

Reply:
Is it possible for two different customer/tenant to create arktos network object with the same name? If yes, we should re-evaluate object store association.

Reply:
in theory it is possible.
tenants won't be the same; tenant name is unique.

Reply:
Arktos provides K8S_POD_TENANT along with K8S_POD_NAMESPACE & K8S_POD_NAME.
Consider to leverage that if you would like to use tenant+ns+pod as key and seek for alternative that does not rely on explicit ani-args mechanism.